src/core/mesh/qgsmeshdataprovidertemporalcapabilities.cpp

```diff
@@ -0,0 +1,146 @@
+#include "qgsmeshdataprovidertemporalcapabilities.h"
+
+
+QgsMeshDataProviderTemporalCapabilities::QgsMeshDataProviderTemporalCapabilities(): QgsDataProviderTemporalCapabilities()
+{}
+
+QgsMeshDatasetIndex QgsMeshDataProviderTemporalCapabilities::datasetIndexTimeInMilliseconds( int group, qint64 timeSinceGlobalReference ) const
+{
+  const QDateTime &groupReference = mGroupsReferenceDateTime[group];
+
+  const qint64 timeSinceGroupeReference =
+    timeSinceGlobalReference - mGlobalReferenceDateTime.msecsTo( groupReference );
+  const QList<qint64> &datasetTimes = mDatasetTimeSinceGroupReference[group];
+  // No dataset or time too little, return invalid index
+  if ( datasetTimes.isEmpty() )
+    return QgsMeshDatasetIndex();
+
+  // If time is less than the time of the first dataset, return the first dataset
+  // Allow to display "static" dataset (as terrain elevation) whatever the time
+  if ( timeSinceGroupeReference < datasetTimes.first() )
+    return QgsMeshDatasetIndex( group, 0 );
+  for ( int i = 1; i < datasetTimes.count(); ++i )
+  {
+    if ( timeSinceGroupeReference < datasetTimes.at( i ) )
+      return QgsMeshDatasetIndex( group, i - 1 );
+  }
+
+  // no time found because time too big --> return the last index;
+  return QgsMeshDatasetIndex( group, datasetTimes.count() - 1 );
+}
+
+QgsMeshDatasetIndex QgsMeshDataProviderTemporalCapabilities::datasetIndex( int group, double timeSinceGlobalReferenceInHours ) const
```

Comment:
I am wondering, wouldn't it make sense to use miliseconds everywhere in qgis, so we do need to do 3600 * 1000 * and / 3600 / 1000 here and there? (just a question, maybe more hassle)...

how it is in WMS-T, how the time is stored?

also if we use consistently ms, we do not need to add "InHours" or "inMiliseconds" to all functions, variables